### PR TITLE
simplify param type

### PR DESCRIPTION
**Pull request by ermalkaleci**
I had this idea to make param type an enum

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -27,32 +27,38 @@ pub struct ServerConfig {
...
 
-#[derive(Deserialize, Debug, Eq, PartialEq)]
-pub struct MethodParam {
-    pub name: String,
-    pub ty: String,
-    pub is_optional: Option<bool>,
-    pub inject: Option<bool>,
+#[derive(Deserialize, Debug, Clone)]
+pub enum ParamType {
+    U32,
```
Comment (xlc) [inline src/config.rs:49]: this is interesting. but I don't want to limit the ParamType. user should be able to specify whatever type and we just ignore ones we don't understand.
maybe somehow make it default to `Type(Strinig)`?

Comment (ermalkaleci) [reply src/config.rs:49]: We can have Any which ignores validation

Comment (ermalkaleci) [reply src/config.rs:50]: @xlc Any type should be enough. We don't need exact type if you're not validating anything

Comment (xlc) [reply src/config.rs:50]: Let's only keep types that we are actually using

Comment (ermalkaleci) [reply src/config.rs:50]: ok removing Any

Comment (xlc) [reply src/config.rs:49]: I mean only keep Any, BlockHash, and BlockNumber

Comment (ermalkaleci) [reply src/config.rs:49]: Ok

**Comment by shaunxw** [comment]
I'm not sure about this. It more like a different way of just saying `with_block_number: index`. I think having full params info instead would be more flexible for future features.

File: config.yml [hunk 2]
```diff
@@ -11,139 +11,90 @@ rpcs:
   methods:
     - method: author_submitExtrinsic
       params:
-        - name: extrinsic
-          ty: Bytes
+        - Any
```
Comment (xlc) [inline config.yml:14]: the type should still be `Bytes`, but serialize into `Any` (or `Unknown` will be a better name)

Comment (ermalkaleci) [reply config.yml]: > the type should still be `Bytes`, but serialize into `Any` (or `Unknown` will be a better name)

Done

File: src/config.rs [hunk 3]
```diff
@@ -27,32 +27,35 @@ pub struct ServerConfig {
...
+#[derive(Clone, Deserialize, Debug)]
+pub enum ParamType {
+    BlockHash,
+    BlockNumber,
+    #[serde(rename = "vec")]
+    Vec(Box<ParamType>),
+    #[serde(rename = "optional")]
+    Optional(Box<ParamType>),
+    #[serde(rename = "at")]
+    At(Box<ParamType>),
```
Comment (shaunxw) [inline src/config.rs:56]: will be good to add docstring as the `At` type implication (optional, inject, block hash or number) is not self-explaining from its name

File: src/server.rs [hunk 4]
```diff
@@ -49,16 +50,21 @@ pub async fn start_server(
...
-                api.clone(),
-                InjectType::BlockHashAt(index),
-            )));
-        } else if let Some(index) = method.inject_block_num() {
-            list.push(Arc::new(InjectParamsMiddleware::new(
-                api.clone(),
-                InjectType::BlockNumberAt(index),
-            )));
+        match method.at() {
+            // TODO: inclue params into middleware and do this inside
```
Comment (shaunxw) [inline src/server.rs:54]: need to do in this PR?

Comment (ermalkaleci) [reply src/server.rs:54]: This is a PR to your branch. I wanted to give this idea and let you finish the feature

Comment (shaunxw) [reply src/server.rs:54]: I'm not fond of this idea very much tbh. I'm not against it merged to my branch if everyone else is happy, but that's all. You should finish your idea/work.

Comment (ermalkaleci) [reply src/server.rs:54]: I'm closing this, let's keep it simple

**Pull request closed without merging**